portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,28 +32,46 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private List<String> comments = new ArrayList<>();
+  private List<String> comments;
   private Gson gson = new Gson();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String text = (String) entity.getProperty("text");
+      comments.add(text);
+    }
+
     String json = gson.toJson(comments);
     response.setContentType("text/html;");
     response.getWriter().println(json);
   }
 
   /**
-   * Retrieve inputted comments and redirect the result to the page.
+   * Retrieve and store inputted comments and redirect the result to the page.
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
-    String text = getParameter(request, "text-input", "");
+    Optional<String> textOptional = getParameter(request, "text-input");
+    String text = textOptional.orElse("");
```

Comment:
Maybe not add the comment if it's not present? Perhaps add some validation on the front end (e.g. disable comment button when there's no text entered?)